clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -66,145 +213,189 @@ class device_vector {
-                                          InputIterator>::iterator_category,
-                                      ::std::random_access_iterator_tag>::value,
-                    InputIterator>::type last)
-      : _alloc(get_default_queue()), _size(::std::distance(first, last)) {
-    _set_capacity_and_alloc();
-    ::std::vector<T> _tmp(first, last);
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                _tmp.begin(), _tmp.end(), this->begin());
-    }
+  device_vector(const device_vector &other, const Allocator &alloc)
+      : _alloc(alloc) {
+    _size = other.size();
+    _capacity = other.capacity();
+    _storage = alloc_traits::allocate(_alloc, _capacity);
+    _construct(other.begin(), other.end());
   }
 
-  template <typename OtherAllocator>
-  device_vector(const device_vector<T, OtherAllocator> &v)
-      : _alloc(get_default_queue()), _storage(v.real_begin()), _size(v.size()),
-        _capacity(v.capacity()) {}
+  device_vector(const device_vector &other)
+      : device_vector(
+            other,
+            alloc_traits::select_on_container_copy_construction(other._alloc)) {
+  }
 
   template <typename OtherAllocator>
-  device_vector(::std::vector<T, OtherAllocator> &v)
-      : _alloc(get_default_queue()), _size(v.size()) {
-    _set_capacity_and_alloc();
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                v.begin(), v.end(), this->begin());
-    }
+  device_vector(const device_vector<T, OtherAllocator> &other,
+                const Allocator &alloc = Allocator(get_default_queue()))
+      : _alloc(alloc) {
+    _size = other.size();
+    _capacity = other.capacity();
+    _storage = alloc_traits::allocate(_alloc, _capacity);
+    _construct(other.begin(), other.end());
   }
 
+  template <typename OtherAllocator>
+  device_vector(const ::std::vector<T, OtherAllocator> &v)
+      : device_vector(v.begin(), v.end()) {}
+
+  template <typename OtherAllocator>
+  device_vector(const ::std::vector<T, OtherAllocator> &v,
+                const Allocator &alloc)
+      : device_vector(v.begin(), v.end(), alloc) {}
+
   template <typename OtherAllocator>
   device_vector &operator=(const ::std::vector<T, OtherAllocator> &v) {
     resize(v.size());
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
+    ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
                 v.begin(), v.end(), begin());
-    }
     return *this;
   }
+
   device_vector &operator=(const device_vector &other) {
     // Copy assignment operator:
+    if constexpr (alloc_traits::propagate_on_container_copy_assignment::value) {
+      clear();
+      alloc_traits::deallocate(_alloc, _storage, _capacity);
+      _capacity = 0;
+      _alloc = other._alloc;
+    }
+    reserve(other.capacity());
     resize(other.size());
```

Comment:
How about use `uninitialized_copy` instead of `resize` and `copy`?

Reply:
Yes, I think this can be improved, but we have a few different scenarios to cover.  
We need special cases for 
1) when the incoming size is <= existing size
`std::copy`, then destroy excess

2) when it isn't but the incoming size is <= _capacity
`std::copy_n` the already constructed elements, `unintialized_device_copy_n` the rest, then set new size

3) when incoming elements exceed current capacity
destroy all existing elements, `uninitialized_device_copy_n` all elements, set size

Reply:
fixed, thanks.